### PR TITLE
feat(dql): add keyword "every" to DQL

### PR DESCRIPTION
**Pull request by EnricoMi**
This adds the keyword `every` to DQL, which picks equidistant nodes. This provides a deterministic sample of the result that is proportional in size. It can be combined with `first`, `offset`, and `after`. Pagination is applied **before** `every`.

This supersedes #8107

Docs PR: None

### CODE REVIEW

File: query/query.go [hunk 1]
```diff
@@ -2286,6 +2296,14 @@ func ProcessGraph(ctx context.Context, sg, parent *SubGraph, rch chan error) {
 		}
 	}
 
+	// We apply Every _after_ pagination, so it refers to the respective page
+	if sg.Params.Every > 1 {
+		if err = sg.applyEvery(ctx); err != nil {
```
Comment (mangalaman93) [inline query/query.go:2301]: we should use := here to reduce the scope of err

**Review by mangalaman93 — CHANGES_REQUESTED** [review]
Thank you for updating the PR. This looks good to me, I have just one minor comment. I am also thinking that we should add test where every and after are used together. Thanks

File: query/query.go (hunk 1 above)
Comment (EnricoMi) [reply query/query.go]: Many other place do not limit the scope of err, so I have reduced the scope in all those other places as well: eebba60020a5ce17be038a23bbb645ac57b5c896

**Comment by EnricoMi** [comment]
> we should add test where every and after are used together.

Added in https://github.com/dgraph-io/dgraph/pull/8976/files#diff-99902c080bca80ed1ad52da34ea4041069d1c64a52317fc17f2285e8170ae6ddR1085

File: query/query.go (hunk 1 above)
Comment (mangalaman93) [reply query/query.go]: Thanks

File: query/query.go [hunk 2]
```diff
@@ -1654,13 +1664,13 @@ func (sg *SubGraph) recursiveFillVars(doneVars map[string]varValue) error {
 		return err
 	}
 	for _, child := range sg.Children {
-		err = child.recursiveFillVars(doneVars)
+		err := child.recursiveFillVars(doneVars)
```
Comment (mangalaman93) [inline query/query.go:1667]: nit: if you are doing this, you can also move the statement inside the if. Same for some of the statements below.

**Review by mangalaman93 — APPROVED** [review]
one minor comment, LGTM otherwise. Is it possible for you to raise a corresponding docs PR on https://github.com/dgraph-io/dgraph-docs

**Comment by EnricoMi** [comment]
I'd love to run tests locally, but the instructions in CONTRIBUTING.md seem outdated and following the instructions from `ci-dgraph-tests.yml` don't work either:

```
cd t; ./t
testutil: "" localhost:9080 localhost:5080
...
Running tests for 121 packages.
Sent 1/121 packages for processing.
Bringing up cluster test-013-1 for package: ../dgraph/docker-compose.yml ...
CLUSTER UP: test-013-1. Package: ../dgraph/docker-compose.yml
Health for test-013-1_alpha6_1 failed: Get "http://localhost:49272/health": read tcp 127.0.0.1:35362->127.0.0.1:49272: read: connection reset by peer. Response: "". Retrying...
Health for test-013-1_alpha6_1 failed: Get "http://localhost:49272/health": dial tcp 127.0.0.1:49272: connect: connection refused. Response: "". Retrying...
Health for test-013-1_alpha6_1 failed: Get "http://localhost:49272/health": dial tcp 127.0.0.1:49272: connect: connection refused. Response: "". Retrying...
...
```

Running `dgraph version` in the `dgraph/dgraph:local` image returns:
```
dgraph: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.34' not found (required by dgraph)
dgraph: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.32' not found (required by dgraph)
```

**Comment by EnricoMi** [comment]
I have reworked the notion of `every` in conjunction with `first`, `after` and `offset`. The `every` is applied before pagination, so that the result of `every` can be paginated, not the paginated is every-ed.

This required to change the way `every` works. It now picks the last uid from `every`-sized batches, so that paginating with `after` works: 2e4f90f2093b7a7b317d58dcfd35775a4a2c6f70

**Comment by EnricoMi** [comment]
I have managed to get tests working by setting `contrib/Dockerfile` to `ubuntu:22.04` to match my dev machine.

File: query/query4_test.go [hunk 3]
```diff
@@ -988,6 +989,246 @@ func TestCascadeSubQuery1(t *testing.T) {
...
+				"name": "Elizabeth"
+			  }
+			]
+		  }
+	}`, js)
+}
+
+func TestHasEveryWithFirstAfter(t *testing.T) {
+	query := `{
+		q(func:has(name), every: 10, first:2, after: 0x36) {
```
Comment (mangalaman93) [inline query/query4_test.go:1186]: Shouldn't this return exactly 2 results?

**Review by mangalaman93 — CHANGES_REQUESTED** [review]
Looks much better too. One question before it is ready to go.

File: query/query4_test.go (hunk 3 above)
Comment (EnricoMi) [reply query/query4_test.go:1186]: It returns exactly two uids, some have multiple `<name>` values.

Comment (mangalaman93) [reply query/query4_test.go:1126]: Did you modify the test now?

Comment (EnricoMi) [reply query/query4_test.go:1126]: Yes, I found another triple in the test data with predicate `<name>`, which changed the sequence of results, so I had to update the expectation: https://github.com/dgraph-io/dgraph/compare/52d436be0c4589528b9dffb1c593fa0a13b1f4de..a8e9948eb6f95dbcb879831a83cff4b465913172

I think that should be the right response, but my code produces some other result.

Comment (EnricoMi) [reply query/query4_test.go:1126]: Is there a way to unit-test `SubGraph.applyEvery` with some test `uidMatrix`?

File: query/query.go [hunk 4]
```diff
@@ -2383,6 +2398,35 @@ func ProcessGraph(ctx context.Context, sg, parent *SubGraph, rch chan error) {
...
+// applies "every" to lists inside uidMatrix
+// every 'every-th' uid is selected from the uid lists,
+// padding uids are considered when skipping uids from next list
+func (sg *SubGraph) applyEvery(ctx context.Context) error {
+	sg.updateUidMatrix()
+
+	offset := 0
+	every := sg.Params.Every
+	for i := 0; i < len(sg.uidMatrix); i++ {
+		uidList := sg.uidMatrix[i].Uids
```
Comment (EnricoMi) [inline query/query.go:2410]: @mangalaman93 assumption here is that uids in `sg.uidMatrix[i].Uids` are sorted ascending order, and all uids in `sg.uidMatrix[i].Uids` are smaller than in `sg.uidMatrix[i+1].Uids`. Is that true?

Comment (harshil-goel) [reply query/query.go:2410]: No, i+1 doesn't have to be bigger than i. We merge these later using MergeSorted. Uids should be in the correct order.

**Review by harshil-goel — APPROVED** [review]

**Comment by mangalaman93** [comment]
@EnricoMi Could you look at the CI failures? 

Also, I only have an incomplete version of this PR https://github.com/dgraph-io/dgraph/pull/8977. If you like, you could take over a complete it. It should be a very similar PR.

File: query/query.go (hunk 4 above)
Comment (EnricoMi) [reply query/query.go:2410]: So the individual `sg.uidMatrix[*].Uids` lists are arbitrary subsets of the entire uid space. This renders the "offset" logic obsolete.

Any idea how to efficiently sample every `every`-th uid deterministically?

Comment (EnricoMi) [reply query/query.go:2410]: Given https://github.com/dgraph-io/dgraph/pull/8977/files#diff-c6a819ee05eb6643352fb889fdd48569cca191565c59c96ec849150a12bad2feR2417, it looks like then notion of having multiple `sg.uidMatrix[*].Uids lists is different to what I presume. What does it mean to have multiple `uidMatrix`?

**Comment by EnricoMi** [comment]
@mangalaman93 happy to finish #8977 once I understand the notion of having multiple `sg.uidMatrix`.

File: query/query.go (hunk 4 above)
Comment (EnricoMi) [reply query/query.go:2410]: I have looked through the documentation of `subGraph` and I think I understand `sd.uidMatrix` now. Updating tests.

File: query/query4_test.go (hunk 3 above)
Comment (EnricoMi) [reply query/query4_test.go]: You were right, it should return only two results, as it selects the `name` property without language annotation per uid.

**Comment by EnricoMi** [comment]
Tests should all pass now, will look into docs PR next.

**Comment by mangalaman93** [comment]
@EnricoMi upgrade tests are failing but don't worry about those. They are failing because the older versions of Dgraph do not support "every" keyword. I will fix those.

File: query/query.go [hunk 5]
```diff
@@ -2383,6 +2398,31 @@ func ProcessGraph(ctx context.Context, sg, parent *SubGraph, rch chan error) {
...
+		for j := 0; j < uidsLen; j++ {
+			// take the last uid of every-sized batches
+			// taking the last rather than the first allows us to paginate via after
+			uids[j] = uidList[(j+1)*every-1]
+		}
+		sg.uidMatrix[i].Uids = uids
+	}
+
+	// Re-merge the UID matrix.
+	sg.DestUIDs = algo.MergeSorted(sg.uidMatrix)
```
Comment (harshil-goel) [inline query/query.go:2422]: Can you also write unit test for applyEvery function? 
I think you need to first MergeSort  the entire uidMatrix and then take every ith element? Right now I don't think we will get the correct result.

File: query/query4_test.go [hunk 6]
```diff
@@ -988,6 +989,142 @@ func TestCascadeSubQuery1(t *testing.T) {
 	}`, js)
 }
 
+func TestHasEvery(t *testing.T) {
+	query := `{
```
Comment (mangalaman93) [inline query/query4_test.go:993]: add this line as the first line of the test in each new test that you have added. This should fix the upgrade tests. Thanks

```
dgraphtest.ShouldSkipTest(t, "071c85b59fdf80ade88f4571e9cf8c867fae1f6a", dc.GetVersion())
```

**Comment by mangalaman93** [comment]
@EnricoMi one more question, what is the use case of "every" word?  Why can't we just use "random"?

**Comment by EnricoMi** [comment]
@mangalaman93 the use case is to partition the result set (the entire graph) into equi-sized partitions with one query to get boundary uids that are then used with `after` to fetch results in parallel: https://discuss.dgraph.io/t/sample-your-result/9055/1 With the `random` key the size of partitions would not be as useful.

Using offset is not an option as it has `O(N)` time complexity, where `after` has constant time: https://discuss.dgraph.io/t/pagination-with-offset-should-scale-like-after/8577

Looking at the code, the `every` keyword seems of `O(N)` complexity as well. Maybe the sampling should happen remotely in `worker.ProcessTaskOverNetwork` to retrieve only the sample of uids (e.g. every 1,000,000th uid): https://github.com/dgraph-io/dgraph/blob/890c1f79b3afeeba7dd66fae54fa74b20b049a25/query/query.go#L2160

**Comment by mangalaman93** [comment]
> Looking at the code, the `every` keyword seems of `O(N)` complexity as well. Maybe the sampling should happen remotely in `worker.ProcessTaskOverNetwork` to retrieve only the sample of uids (e.g. every 1,000,000th uid):

that does seem like an interesting idea. I just wonder whether it is feasible to do given that results may get filtered out later sometimes. If you want, you can look into it. Thanks for the explanation for adding 'every' to the query language.

**Comment by EnricoMi** [comment]
> that does seem like an interesting idea

I may need some pointer to the code where that `worker.ProcessTaskOverNetwork` is being executed.

**Comment by EnricoMi** [comment]
@mangalaman93 ping

**Pull request closed without merging**